test/management/attack-protection.tests.js

```diff
@@ -0,0 +1,208 @@
+var expect = require('chai').expect;
+var nock = require('nock');
+
+var SRC_DIR = '../../src';
+var API_URL = 'https://tenant.auth0.com';
+
+var AttackProtectionManager = require(SRC_DIR + '/management/AttackProtectionManager');
+var ArgumentError = require('rest-facade').ArgumentError;
+
+describe('AttackProtectionManager', function() {
+  before(function() {
+    this.token = 'TOKEN';
+    this.attackProtection = new AttackProtectionManager({
+      headers: { authorization: 'Bearer ' + this.token },
+      baseUrl: API_URL
+    });
+  });
+
+  describe('instance', function() {
+    var methods = ['getBruteforceConfig', 'updateBruteforceConfig'];
+
+    methods.forEach(function(method) {
+      it('should have a ' + method + ' method', function() {
+        expect(this.attackProtection[method]).to.exist.to.be.an.instanceOf(Function);
+      });
+    });
+  });
+
+  describe('#constructor', function() {
+    it('should error when no options are provided', function() {
+      expect(AttackProtectionManager).to.throw(ArgumentError, 'Must provide manager options');
+    });
+
+    it('should throw an error when no base URL is provided', function() {
+      var client = AttackProtectionManager.bind(null, {});
+
+      expect(client).to.throw(ArgumentError, 'Must provide a base URL for the API');
+    });
+
+    it('should throw an error when the base URL is invalid', function() {
+      var client = AttackProtectionManager.bind(null, { baseUrl: '' });
+
+      expect(client).to.throw(ArgumentError, 'The provided base URL is invalid');
+    });
+  });
+
+  describe('#getBruteforceConfig', function() {
+    var data = {
+      enabled: true,
+      shields: ['user_notification', 'block'],
+      mode: 'count_per_identifier_and_ip',
+      allowlist: ['1.1.2.2'],
+      max_attempts: 100
+    };
+
+    beforeEach(function() {
+      this.request = nock(API_URL)
+        .get('/attack-protection/brute-force-protection')
+        .reply(200);
+    });
+
+    it('should accept a callback', function(done) {
+      this.attackProtection.getBruteforceConfig({}, function() {
+        done();
+      });
+    });
+
+    it('should return a promise if no callback is given', function(done) {
+      this.attackProtection
+        .getBruteforceConfig()
+        .then(done.bind(null, null))
+        .catch(done.bind(null, null));
+    });
+
+    it('should pass any errors to the promise catch handler', function(done) {
+      nock.cleanAll();
+
+      var request = nock(API_URL)
+        .get('/attack-protection/brute-force-protection')
+        .reply(500);
+
+      this.attackProtection.getBruteforceConfig().catch(function(err) {
+        expect(err).to.exist;
+
+        done();
+      });
+    });
+
+    it('should pass the body of the response to the "then" handler', function(done) {
+      this.attackProtection.getBruteforceConfig().then(function(bruteforceConfig) {
+        expect(bruteforceConfig.max_attempts).to.equal(data.max_attempts);
```

Comment:
I wonder if we could do something with `deep equals` here. Note that the data we use in this script doesn't really have to match the shape of the actual data from the BFP endpoint.

Reply:
👍